### ISSUE TITLE
Enhancement: optional "face=n" and/or "specular=xxxx" and/or "normal=yyyy" in .themes

### ISSUE THREAD

**Issue by RestlessSwords** [OpenCollarTeam/OpenCollar#133]
What version of OpenCollar are you using?
- 6.x or 7.x

What behavior did you expect?
- sometimes only want to change a specific face on a particular named prim(s)
- sometimes want to add a NormalMap (bump texture) and/or a SpecularMap (shininess texture)

What behavior did you see instead?
- all faces always changed
- new .meshthemes and oc_themes only works for ONE prim, and many collars have many prims

What steps does someone need to take to reproduce the problem?
- any collar using .themes and oc_themes to change prim colors and/or textures

the new (optional) parameter in .themes is illustrated in example below:
```
Example .themes notecard:
[ Leather Flat ]
Band~face=0~ChokerLeather~<0.99, 0.99, 0.99>~normal=0~specular=0
Band~face=1~ChokerMetalBlack~<0.99, 0.99, 0.99>~normal=0~specular=0
Band~face=2~ChokerMetalRed~<0.99, 0.99, 0.99>~normal=0~specular=0

[ Normal Specular ]
Band~texture=ChokerLeather~color=<0.99, 0.99, 0.99>~normal=95a5a0d7-dc08-f128-1f64-b109849b9d9b~specular=e3e4b750-1ee7-30db-381e-9ef307e218b9
Band~face=1~ChokerMetalBlack~<0.99, 0.99, 0.99>~normal=ChokerLeatherNormal~specular=ChokerLeatherSpecular
Band~face=2~normal=ChokerLeatherNormal~specular=ChokerLeatherSpecular~ChokerMetalRed~<0.99, 0.99, 0.99>
```

see pull-request #135 
[ (https://github.com/OpenCollarTeam/OpenCollar/pull/135) ]

As this example shows:
- the (optional) "face" parameter needs to be before any texture or color parameters on the line
- the (optional) "specular" and "normal" parameters can be anywhere on the line
- the "specular" and "normal" values can be full texture UUIDs or shortnames from the .textures file

**Comment by RestlessSwords** [comment]
the GitHub removed the "~" (tilde) before and after the words it stroked-through

**Comment by Medea-Destiny** [comment]
This looks like a good set of enhancements. 

My concern is that this approach dilutes the purpose of defining elements. My feeling is that it would be better to keep a single definition for each element rather than a per-face definition, and allow the collar creators to define per-face elements. So for example, for the above example for .themes, you might define in the object desc:

Band~0,MetalBlack~1,MetalRed~2

... where desc is parsed in the format element~face,element~face

and then stored in a strided list in a format something like:

element1, link~face|link~face|link~face, element2,link~face etc.

I believe this as an enhancement to oc_themes would make a separate oc_meshthemes script unnecessary, which will save user confusion, while retaining a simple per-element setting.

**Comment by RestlessSwords** [comment]
Medea:
That's an interesting idea.  
But, for better or worse,  I tried to keep the changes consistent with the current .themes organization.
- Current .themes has one line per (entire) prim element being textured/colored (eg "Band")
- My approach adds optional "face=n" to make the line apply to a prim face (when they are different)
- so, if previously had 6 lines for 6 prim elements, *might* now have 6 lines for 3 prims each of 2 faces.
- i.e. the (optionally) specified face is now essentially treated as another independent prim/element

*AND* you don't want the face textures (MetalBlack, etc) fixed in the prims description
- I believe the whole point of having .themes is so those can be selected and dynamically varied
- (unless I am mis-understanding your proposal)

**Comment by RestlessSwords** [comment]
Medea:
To clarify my currently implemented approach:
- assume a collar made up of 2 prims (native or mesh, does not matter)
- prim-1 mostly using one set of textures, but prim-1 having face-4 and face-6 with different texture set
- prim-2 mostly using another set of textures, but prim-2 having face-3 and face-5 with different textures
The .themes would look like:
```
[ themeXX ]
prim1~texture1
prim1~face=4~texture14
prim1~face=6~texture16
prim2~texture2
prim2~face=3~texture13
prim2~face=5~texture15
[ themeYYY ]
.... more stuff
```
How would your approach look ?

**Comment by Medea-Destiny** [comment]
@RestlessSwords 

My thought is that the elements system allows us to divide the build up according to what the parts of the collar represent in terms of material, texture, etc. Certain parts of the collar will represent particular elements of the design -- the metal rings, the leather straps, the buckle, the lock, etc. These elements are the basic unit of the design -- we might want the ring to be chrome metal, we might want it to be red plastic, we might want it to be rubberised. Back in the ancient days of prim-only builds, this tended to be a simple enough thing, because a prim was virtually never going to represent anything other than a whole or a part of an element.

With a mesh build, the prim as a unit of design no longer makes sense. What I believe still makes sense is the concept of a design element, which would be defined by a shared texture (or textures -- diffuse, specular, and normal). We might have a dozen rivets in the design which are individual faces of multiple different objects, but they all share a material property -- i.e. they are common elements of the design. When changing aspects of the appearance, either manually or as part of a theme, it makes sense to do so by these elements. If we start applying different texture maps to different instances of an element, I think we lose that utility.

From the perspective of design, an element is similar to a material slot in Blender. We might not be talking about parts that share a uv map the way the above example of rivets do, even -- for example think of the case of a collar where all the metallic parts use a single texture atlas. However they will share a texture, specular properties, tinting, etc.

In your above design, if I understand it rightly, it might look something like this:

```
prim 1 description: metals~1,metals~2,metals~3,lock~4,metals~5,clasp~6
prim 2 description: trim~1,trim~2,hooks~3,trim~4,hooks~5
prim 3 description: band~1,band~2,band~3,metals~4
... where each prim is given a definition element~face,element~face,element~face

```
Actually it might make more sense to have the first element defined inherited by all faces that aren't specifically defined, so more like:

`prim 1 description: metals,lock~4,clasp~6`

The descriptions of the link set are looped through and a strided list of elements created, which in the above case would look something like:

`g_lElements=["metals","1~1|1~2|1~3|1~5|3~4","lock","1~4","clasp","1~6","trim","2~1|2~2|2~4","hooks","2~3|2~5","band","3~1|3~2|3~3"];`

These distinct parts -- metals, lock, clasp, trim, hooks and band -- can be defined individually through a texturing menu, or as a whole via the .themes notecard, something like:

```
[theme 1]
band~texture=BlackLeather~color=<1,1,1>~normal=0~specular=0
metals~texture=ShinyChrome~color=<0.9,0.9,1>~normal=0~specular=(uuid)
hook~texture=brushed_brass~color=<1,1,0.9>~normal=(uuid)~specular=TEXTURE_BLANK~specCol=<0.5,0.7.1>
(etc)

[theme 2]
band~texture=EmbossedVelvet~normal=EVNorm~specular=0
(etc)
```
There are backwards compatibility issues to consider, as with the change to oc_meshthemes, but I think this offers the most clarity and convenience going forwards.

**Comment by RestlessSwords** [comment]
@Medea-Destiny 
1) I think I understand your proposal, but I am concerned it starts to create a whole new level of indirection that buries some of the information about textures.  for example your new high-level item "metals" is now in the description of each of many of the collar prims in your example.
2) as well I have metal rivets in several places in my collars -- but they are defined in the 3D builder (blender) and when the DAE file is exported and imported into SL, those items already have their textures sub-areas inside the large prim textures exported out of blender, and there is no easy way inside SL to extract them and use some new "metals" texture since the exported/imported model has definitions of the position/size/repeats etc of where their piece of texture is expected to be.

In summary, I think the way I have done it easily allows a prim to (optionally) have a face which we can chose to color etc independent of the other prim faces, and does it in a way that introduces ZERO incompatibilities with old .theme notecards -- which I believe is absolutely critical, we cannot have a new oc_themes version that wont work with all old collars.

**Comment by RestlessSwords** [comment]
I wrote a semi-detailed reply, saying why I thing the way I did it works
best (especially considering requirement for backward compatibility)

**Comment by Medea-Destiny** [comment]
@RestlessSwords 

1. This isn't really a "new level", this is how the collar appearance system worked for a long time. For example this is how we hide particular elements of the collar, such as the bell.  I don't think "indirection" is a fair term to use here. To the contrary, it's a way of clarifying to the user what parts of the collar are being referred to.

2. One or other of us isn't quite understanding the other here a bit, and I don't know who :)
There's no need to "extract" anything here. Consider a leather collar with metal rivets. You may have both the leather area and the metal area sharing a single texture, with the leather taking up some part of the UV space, and the metal area taking up other parts of the UV space.  That's fine. Suppose you render  out two versions of this texture, one with chrome metal and black leather, the other with brass metal and white leather. Assuming you have given the leather area and the metal area different materials in Blender, and thus they import into SL as having different faces, you don't need to worry about position/repeats. Apply texture 1 to the face that has leather on it and texture 2 to the face that has metal on it, and you'll have black leather with brass metals. 
It may be that in your design you opt to have both the leather and the metal as a single face. In that case your element name might be something like "collar base" and you'd swap the atlased texture for both at the same time. So I'm not really sure what your objection is here.

Backwards compatibility is an issue, but not one that can't be dealt with using a little care. Where faces aren't specified, then we can treat the line as previously anyway. This could all be a functional add-on, with notecard lines flagged in some way to indicate they use the new standard.

**Comment by RestlessSwords** [comment]
1. yes, using the description field of a SINGLE prim element, like a bell, to put additional information about it is not a problem.  but in your previous example "metals" was necessary to put in the description of multiple DIFFERENT prim elements (your prim-1 and prim-3 and possible others).  that is what bothered me about that
2. what you describe is the way one of my collars currently works, the original collar+metal texture I duplicated and renamed Band and Metal, and then have colored versions of each of them. 

Overall, it just seems that there are 2 different approaches.  I put the face specific decisions into the .themes whereas you would put them into the prim descriptions.  the end result should be functional both ways, it just seems that your approach requires additional code since there *IS* a new level of abstraction introduced by your named sub-Elements in the prim descriptions.  I think we agree to disagree which is easiest.  In total, adding "face" support was 10-15 lines of code I changed, and adding "Specular" and "Normal" were each about 10-15 lines.  I believe adding the code to get the prim description sub-Element, code to store them, and code to use them in the existing operations (texture, color) and creating new code operations (specular, normal) that use them, would be a lot more extensive.
- and collar builders/modifies would need to create and maintain both the .themes notecard *AND* prim descriptions.

So I believe the approach I implemented is simpler (in concept, implementation, and maintenance) although I understand that you think otherwise.

**Comment by Medea-Destiny** [comment]
@RestlessSwords 

I agree your approach is simpler, I just think that mine is better ;)

I think this "new level of abstraction" is actually more in keeping with the old method of abstraction, just updated to allow for multiple faces. Collar builders would indeed need to maintain both .themes and prim descriptions -- but they should be doing that anyway, and my method makes for simpler and cleaner .themes notecards, and indeed from a designer standpoint it means that .themes could very easily be shared across multiple variant designs, rather than .themes being specific to a particular arrangement of links/faces.

In short I'd put it like this. Your approach is there, simple to implement, and waiting on a pull. That's a major bonus. Mine is more flexible and future-proof, but requires a more radical re-write of the themes script. The question is -- do we go for the quick solution that works, and risk having to change things again a little while later, or do we hold and go for a more complicated solution that (at least in my view) is a better one?

**Comment by RestlessSwords** [comment]
LAUGHS -- *Better*, like *Beauty* is very much in the eye of the beholder.

**Comment by nirea** [comment]
I'm leaning away from the concept of "elements" that require prim descriptions (even though I know I introduced it way back when).  @Medea-Destiny does your proposal require such descriptions?

A couple more thoughts:
1. If we're revamping theme scripts, it'd be nice to be able to unify oc_themes and oc_meshthemes.  Would both of the solutions presented here work on a single-link collar that just uses faces?
2. I'd like to avoid imposing constraints on designers that they need to chop up their meshes a certain way to give users a good re-theming experience.  It seems like using prim descriptions would impose such a constraint though.

Lastly, and this may be a lost cause, but I hate the markup format for the .themes notecard.  It requires a ton of fiddly code to parse, and these cards aren't edited so often that we need a really pretty markup format.  So I have a mild preference for putting improvements into oc_meshthemes until we can deprecate oc_themes.

**Comment by RestlessSwords** [comment]
@nirea 
My modified oc_themes and its .themes notecard works fine on a one-prim collar with many faces (mesh or prim), and works fine on a many prim collar with many faces (mesh or prims).

I purposefully keep the changes compatible with all existing .theme notecards so all old stuff still works.

**Comment by zontreck** [comment]
Please see issue #198 as it might resolve the issue if implemented
